forge-game/src/main/java/forge/game/spellability/AbilityManaPart.java

```diff
@@ -168,14 +186,17 @@ public final void produceMana(final String produced, final Player player, SpellA
         // add the mana produced to the mana pool
         manaPool.add(this.lastManaProduced);
 
-        // Run triggers
-        final Map<AbilityKey, Object> runParams = AbilityKey.mapFromCard(source);
-        runParams.put(AbilityKey.Player, player);
-        runParams.put(AbilityKey.Produced, afterReplace);
-        runParams.put(AbilityKey.AbilityMana, root);
-        runParams.put(AbilityKey.Activator, root == null ? null : root.getActivatingPlayer());
+        if (runTriggers) {
```

Comment:
Thanks, but not good enough yet :/

The problem is if other players control _Caged Sun_ with its weird trigger, because it has Player$ condition, so skipping isn't acceptable I think

Not sure for best approach, maybe new TriggerType or whatever Hanmac suggests :P